utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,212 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFIndex, CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::{
+        kCFStringEncodingUTF8, CFStringGetCString, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    },
+    timezone,
+};
+use libc::c_char;
+use std::ffi::{CStr, CString};
+
+use crate::RetrievalError;
+
+/// Helps to get string, when a string cannot be obtained in O(1) time and space directly from the pointer
+/// A buffer is created of size `length + 1`. After copying the string into the buffer, the null characters `\0`
+/// are removed from the string, which helps it converting to icu4x locale in tests
+fn get_string(ptr: CFStringRef) -> Result<String, RetrievalError> {
```

Comment:
please put the whole logic with both branches of the if block in this method. The case where you go through`CStr` is duplicated multiple times below, and should be part of `get_string`.